elkscmd/man/man1/urlget.1
.TH URLGET 1
.SH NAME
urlget, ftpget, httpget - retrieve a file from the internet to stdout
.SH SYNOPSIS
\fBurlget\fR [\fB-h\fR] [\fB-d\fR] [\fB-p\fR] [\fBtcp:\fR|\fBftp:\fR|\fBhttp:\fR]url

\fBhttpget\fR [\fB-h\fR] [\fB-d\fR] [\fB-p\fR] [http://]host[:port] path

\fBftpget\fR [ftp://]host[:port] [\fB-v\fR] path[/] [user [pass]]

\fBftpput\fR [ftp://]host path[:port] remotepath [user [pass]] localpath
.br
.SH DESCRIPTION
.B Urlget
gets a file specified by a URL and copies it to standard output.

.B Ftpget
similarly gets a file from a conventional ftp server, a login name
and password can be specified.
If a slash is appended to the path, a directory listing is performed
instead of a file transfer.

.B Ftpput
transfers a file to an ftp server, a login name and password can be specified.

.B Httpget
is similar to  urlget,  but  the  host  and  path  are  specified
separately without a scheme, as with ftpget.
.SS OPTIONS
.TP 5
.B \-h
show the status line and MIME header from the server.
.TP 5
.B \-d
discard the file body.
.TP 5
.B \-p
use POST method, otherwise use GET.
.TP 5
.B \-v
For
.B ftpget,
long directory listing or verbose file transfer.
.TP 5
.B \-U
Act as urlget.
.TP 5
.B \-F
Act as ftpget.
.TP 5
.B \-P
Act as ftpput.
.TP 5
.B \-H
Act as httpget.
.SH EXAMPLES
.TP 20
.B urlget http://minix1.woodhull.com/pub/contrib/file.tar.Z > file.tar.Z
    # Download file.tar.Z
.TP 20
.B ftpget minix1.woodhull.com /pub/contrib/README.txt > README.txt
.BR
    # Get a file from an anonymous ftp server
.TP 20
.B httpget -dh minix1.woodhull.com/index.html
.BR
    # Inspect the header of a web page
.TP 20
.SH NOTES
These commands execute the  same  binary  under  different  names.  These
commands  provide  a  lightweight  non-interactive command-line method of
downloading a file or inspecting the status of a web page. Data retrieved
are written to standard output.
.SH AUTHOR
     Michael Temari <Michael@TemWare.Com>


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/inet/urlget/urlget.c
/* urlget.c Copyright 2000 by Michael Temari All Rights Reserved
 *
 * 04/05/2000 Michael Temari <Michael@TemWare.Com>
 * 09/29/2001 Ported to ELKS(and linux) (Harry Kalogiroy <harkal@rainbow.cs.unipi.gr>)
 * 10/16/2021 Added ftpput, verbose option and improved error handling <helge@skrivervik.com>
 *
 * Implements HTTP POST & GET, FTP GET,PUT and directory listings, TCP GET (typical via 
 * netcat at the other end.
 * 
 * Assumes hard links to the appropriate names - ftpget, ftpput, urlget, tcpget, tcpput.
 * Optons (http only):
 * 	-h -- include header in output stream (httpget)
 * 	-d -- discard data (httpget)
 * 	-p -- post instead of get (httpget), data to post (ascii/UTF) is appended to the URL (after a '?').
 *	-v -- verbose file listing & error reporting, progress meter (ftpput/ftpget)
 * 3/7/2025 Added -U, -F, -P, -H for urlget, ftpget, ftpput, httpget
 */


#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <unistd.h>
#include <time.h>


#define _PROTOTYPE(a,b) a b

_PROTOTYPE(char *unesc, (char *s));
_PROTOTYPE(void encode64, (char **pp, char *s));
_PROTOTYPE(int net_connect, (char *host, int port));
_PROTOTYPE(void net_close, (int fd, int errflag));
_PROTOTYPE(char *auth, (char *user, char *pass));
_PROTOTYPE(int skipit, (char *buf, int len, int *skip));
_PROTOTYPE(int httpget, (char *host, int port, char *user, char *pass, char *path, int headers, int discard, int post));
_PROTOTYPE(void ftppasv, (char *reply));
_PROTOTYPE(int ftpreply, (FILE *fpr));
_PROTOTYPE(int ftpcmd, (FILE *fpw, FILE *fpr, char *cmd, char *arg));
_PROTOTYPE(int ftpio, (char *host, int port, char *user, char *pass, char *path, int type, int verbose));
_PROTOTYPE(int tcpget, (char *host, int port, char *user, char *pass, char *path));
_PROTOTYPE(int main, (int argc, char *argv[]));

char ftpphost[15+1];
unsigned int ftppport;
int use_stdin = 0;
char *progname;

/* operating modes */
#define	SCHEME_URLGET   1
#define	SCHEME_FTPGET	2
#define	SCHEME_FTPPUT	3
#define	SCHEME_HTTPGET	4

#define TRANS_DEBUG	0	/* for debug dumps */

char buffer[4096];

/*
 * urlget  [-hp]  [tcp:|ftp:|http:]//url
 * httpget [-hpd] http://host[:port] path
 * ftpget  [-v]   ftp://host[:port] path[/] [user [pass]]
 * ftpput         ftp://host[:port] remotepath [user [pass]] localpath
 */
int usage(void)
{
    fprintf(stderr,
     "Usage: %s [-UFPHhpdv] [http:|ftp:|tcp://host[:port][url] [path][/] [user [pass]]\n",
     progname);
    return -1;
}

char *unesc(char *s) {
   char *p, *p2;
   unsigned char c;

   p = s;
   p2 = s;
   while (*p) {
   	if (*p != '%') {
   		*p2++ = *p++;
   		continue;
   	}
   	p++;
   	if (*p == '%') {
   		*p2++ = *p++;
   		continue;
   	}
   	if (*p >= '0' && *p <= '9') c = *p++ - '0'; else
   	if (*p >= 'a' && *p <= 'f') c = *p++ - 'a' + 10; else
   	if (*p >= 'A' && *p <= 'F') c = *p++ - 'A' + 10; else
   		break;
   	if (*p >= '0' && *p <= '9') c = c << 4 | (*p++ - '0'); else
   	if (*p >= 'a' && *p <= 'f') c = c << 4 | ((*p++ - 'a') + 10); else
   	if (*p >= 'A' && *p <= 'F') c = c << 4 | ((*p++ - 'A') + 10); else
   		break;
   	*p2++ = c;
   }
   *p2 = '\0';
   return(s);
}

void encode64(pp, s)
char **pp;
char *s;
{
char *p;
char c[3];
int i;
int len;
static char e64[64] = {
	'A','B','C','D','E','F','G','H','I','J','K','L','M',
	'N','O','P','Q','R','S','T','U','V','W','X','Y','Z',
	'a','b','c','d','e','f','g','h','i','j','k','l','m',
	'n','o','p','q','r','s','t','u','v','w','x','y','z',
	'0','1','2','3','4','5','6','7','8','9','+','/' };

   p = *pp;
   len = strlen(s);
   for (i=0; i < len; i += 3) {
   	c[0] = *s++;
   	c[1] = *s++;
   	c[2] = *s++;
   	*p++ = e64[  c[0] >> 2];
   	*p++ = e64[((c[0] << 4) & 0x30) | ((c[1] >> 4) & 0x0f)];
   	*p++ = e64[((c[1] << 2) & 0x3c) | ((c[2] >> 6) & 0x03)];
   	*p++ = e64[  c[2]       & 0x3f];
   }
   if (i == len+1)
   	p[-1] = '=';
   else
   if (i == len+2) {
   	p[-1] = '=';
   	p[-2] = '=';
   }
   *p = '\0';
   *pp = p;
   return;
}

char *auth(user, pass)
char *user;
char *pass;
{
static char a[128];
char up[128];
char *p;

   strcpy(a, "BASIC ");
   p = a + 6;
   sprintf(up, "%s:%s", user, pass);
   encode64(&p, up);

   return(a);
}

/*
 * Find the occurence of 2 CR-LF pairs in a row, which indicates the end
 * of the header, start of content.
 */
int skipit(char *buf, int len, int *skip) {

   static int lf = 0;
   static int crlf = 0;
   char *p = buf;

   while (--len >= 0) {
   	if ((crlf == 0 || crlf == 2) && *p == '\r')
   		crlf++;
   	else
   	if ((crlf == 1 || crlf == 3) && *p == '\n')
   		crlf++;
   	else
   		crlf = 0;
   	if (*p == '\n')
   		lf++;
   	else
   		lf = 0;
   	if (crlf == 4 || lf == 2) {
   		*skip = 0;
   		return(len);
   	}
   	p++;
   }

   return(0);
}

/* Display an error message*/
void errmsg(const char *msg, ...)
{
    va_list args;

    va_start(args, msg);
    fprintf(stderr, "%s: ", progname);
    vfprintf(stderr, msg, args);
    fprintf(stderr, "\n");
    va_end(args);
}

int httpget(char *host, int port, char *user, char *pass, char *path, 
	int headers, int discard, int post) {

   int fd, skip, s, s2;
   int len = 0;
   char *a, *qs;

   if (port == 0)
   	port = 80;

   fd = net_connect(host, port);
   if (fd < 0) {
	perror(host);
	return -1;
   }

   if (post) {
   	qs = strrchr(path, '?');
   	if (qs != (char *)NULL) {
   		*qs++ = '\0';
   		len = strlen(qs);
	}
   }

   if (post && len > 0)
	write(fd, "POST ", 5);
   else
	write(fd, "GET ", 4);
   write(fd, path, strlen(path));
   write(fd, " HTTP/1.0\r\n", 11);
   write(fd, "User-Agent: urlget\r\n", 20);
   write(fd, "Connection: Close\r\n", 19);
   if (*user) {
   	write(fd, "Authorization: ", 15);
   	a = auth(user, pass);
   	write(fd, a, strlen(a));
   	write(fd, "\r\n", 2);
   }
   if (post && len > 0) {
   	sprintf(buffer, "Content-Length: %u\r\n", len);
   	write(fd, buffer, strlen(buffer));
   }
   write(fd, "Host: ", 6);
   write(fd, host, strlen(host));
   write(fd, "\r\n", 2);
   write(fd, "\r\n", 2);
   if (post && len > 0)
   	write(fd, qs, len);

   skip = 1;
   while ((s = read(fd, buffer, sizeof(buffer))) > 0) {
   	if (skip) {
   		s2 = skipit(buffer, s, &skip);
   		if (headers)
   			write(1, buffer, s - s2);
   	} else
   		s2 = s;
   	if (s2 && !discard)
		if (write(1, &buffer[s - s2], s2) != s2) {
			perror("write");
			return(-1);
		}
   }

   net_close(fd, 0);	/* send FIN on close */

   return 0;
}

void ftppasv(char *reply) {
char *p;
unsigned char n[6];
int i;

   ftppport = 0;

   p = reply;
   while (*p && *p != '(') p++;
   if (!*p) return;
   p++;
   i = 0;
   while (1) {
   	n[i++] = atoi(p);
   	if (i == 6) break;
   	p = strchr(p, ',');
   	if (p == (char *)NULL) return;
   	p++;
   }
   sprintf(ftpphost, "%d.%d.%d.%d", n[0], n[1], n[2], n[3]);
   ftppport = n[4] * 256 + n[5];
   return;
}

int ftpreply(FILE *fpr) {
static char reply[256];
int s;
char code[4];
int ft;

   do {
   	ft = 1;
   	do {
   		if (fgets(reply, sizeof(reply), fpr) == (char *)NULL)
   			return(-1);
   		if (ft) {
   			ft = 0;
   			strncpy(code, reply, 3);
   			code[3] = '\0';
   		}
   	} while (strncmp(reply, code, 3) || reply[3] == '-');
   	s = atoi(code);
   } while (s < 200 && s != 125 && s != 150);
   if (s == 227) {
     ftppasv(reply);
     return 227;
   }
   return s;
}

int ftpcmd( FILE *fpw, FILE *fpr, char *cmd, char *arg)
{
   int s = 0;
   fprintf(fpw, "%s%s%s\r\n", cmd, *arg ? " " : "", arg);
   fflush(fpw);
   s=ftpreply(fpr); 
   return s;
}

int ftpio(char *host, int port, char *user, char *pass, char *path, int type, int verbose) {
   int fd;
   int fd2;
   FILE *fpr;
   FILE *fpw;
   int s;
   int s2 = 0;
   char *p;
   char *p2;
   char typec[2], *list;
   int infile = -1;

   if (port == 0)
   	port = 21;

   if (type == '\0')
   	type = 'i';

   if (type == 'S') {	/* ftpput */
	if (!use_stdin && (infile = open(path, O_RDONLY)) < 0) {
		perror(path);
		return(-1);
	}
	if (use_stdin) infile = 0;
	type = 'i';
   }
   fd = net_connect(host, port);
   if (fd < 0) {
	perror(host);
	return -1;
   }
   fpr = fdopen(fd, "r");
   fpw = fdopen(fd, "w");

   s = ftpreply(fpr);	/* Get server ID message, should be 220 status. */
   if (s / 100 != 2) {
	errmsg("Connect error: %d", s);
	goto error;
   }
   s = ftpcmd(fpw, fpr, "USER", *user ? user : "ftp");
   if (s / 100 == 3)
   	s = ftpcmd(fpw, fpr, "PASS", *pass ? pass : "urlget@x.com");

   if (s / 100 != 2) {
	errmsg("Authentication failed: %d\n", s);
	goto error;
   }

   p = path;
   if (*p == '/') p++;
   while ((p2 = strchr(p, '/')) != (char *)NULL) {
   	*p2++ = '\0';
   	s = ftpcmd(fpw, fpr, "CWD", unesc(p));
	if ((s/100) != 2) {		/* 250 = success */
		if (infile >= 0) {		/* ftpput: Try to create the directory */
   			s = ftpcmd(fpw, fpr, "MKD", unesc(p));
			if ((s/100) != 2) {	/* 257 = success */
				errmsg("Create directory %s failed, error %d", p, s);
				goto error;
			} 
			if (verbose) errmsg("Destination directory created: %s", p);
   			s = ftpcmd(fpw, fpr, "CWD", unesc(p)); /* assume success */

		} else {
			errmsg("Remote change directory failed: %s -- status %d", p, s);
			goto error;	/* assume this is fatal */
		}
	}
   	p = p2;
   }

   sprintf(typec, "%c", type == 'd' ? 'A' : type);
   s = ftpcmd(fpw, fpr, "TYPE", typec);
   if (s / 100 != 2) {
	errmsg("Type error: %d", s);
	goto error;
   }
   if (strlen(p) == 0) type = 'd'; 	/* last char is '/', it's a directory, list files */
   ftppport=0; 				/* to check if retrieved below */

   s = ftpcmd(fpw, fpr, "PASV", "");
 
   if (ftppport==0) {			/* set in ftpcmd */
	errmsg("Error listing directory: %d", s);
	goto error; 
   }
   fd2 = net_connect(ftpphost, ftppport);
   if (fd2 < 0) {	
	perror(ftpphost);
	goto error;
   }
   if (verbose) 
	list = "LIST"; 
   else 
	list = "NLST";
   if (infile < 0) { 		/* ftpget */
	s = ftpcmd(fpw, fpr, type == 'd' ? list : "RETR", unesc(p));
   	if ((s/100) != 1) {
		errmsg("Cannot open remote file: %d", s);
		goto error;
   	}
	while ((s = read(fd2, buffer, sizeof(buffer))) > 0) {
		if (verbose) fprintf(stderr, ".");
   		s2 = write(1, buffer, s);
   		if (s2 != s) { s = -1; break; }	/* ERROR */
	}
	if (verbose) fprintf(stderr, ".\n");
	if (s < 0) {
		errmsg("File write error");
		if (s2 == -1)
			perror("");
		else
			fprintf(stderr,".\n");
		net_close(fd2, 1);	/* send RST on close */
		/* should delete destination file: stdout */
		s = -1;
		goto error;
	}
	s = ftpreply(fpr);	/* get 'Transfer Complete' message */
	if (verbose)
		errmsg("Exit status %d", s);
	s = 0;

   } else { 		/* ftpput */
	s = ftpcmd(fpw, fpr, "STOR", unesc(p));
   	if (s / 100 != 1) {
		errmsg("Cannot open destination file: %d", s);
		/* 553 is the most common error */
		if (s == 553) errmsg("Permission denied");
		goto error;
   	}

	while ((s2 = read(infile, buffer, sizeof(buffer))) > 0) {
		if (verbose) fprintf(stderr, ".");
		s = write(fd2, buffer, s2);
   		if (s2 != s) { s = -1; break; }
	}
	if (verbose) fprintf(stderr, ".\n");
	if (s < 0) {
		/* The server will RST the data channel on write errors,
		 * no need to do any housekeeping at this end. */
		errmsg("Network write error, transfer aborted");
		s2 = ftpcmd(fpw, fpr, "DELE", unesc(p)); /* delete remote file */
		if (verbose)
			errmsg("Cleaning up - removing %s, status %d", p, s);
	} else
		s = 0;
	close(infile);
   }
   net_close(fd2, s);	/* s == 0? FIN: RST */

error:
   (void) ftpcmd(fpw, fpr, "QUIT", "");

   /* flush buffered output, then close descriptor first so that FIN/RST works*/
   fflush(fpw);
   net_close(fd, s);	/* s == 0? FIN: RST */

   fclose(fpr);		/* associated fd already closed above*/
   fclose(fpw);

   return(s == 0 ? 0 : -1);
}


int tcpget(char *host, int port, char *user, char *pass, char *path) {

   int fd;
   int s;
   int s2;

   if (port == 0) {
	errmsg("No port specified");
	return -1;
   }

   fd = net_connect(host, port);
   if (fd < 0) {
	perror(host);
	return -1;
   }
   if (*path == '/')
   	path++;

   write(fd, path, strlen(path));
   write(fd, "\n", 1);
   while ((s = read(fd, buffer, sizeof(buffer))) > 0) {
   	s2 = write(1, buffer, s);
	if (s2 != s) {
	    net_close(fd, 1); /* send RST*/
	    return -1;
	}
   }
   net_close(fd, 0);	/* send FIN*/
   return 0;
}

int main(int argc, char **argv) {

   char *url, scheme;
   char user[64], pass[64], host[64];
   int port, s;
   int type = 'i';	/* default ftp type */
   char *path, *ps, *p, *at;
   int opt_d = 0, opt_h = 0, opt_p = 0, opt_v = 0;

   if ((progname = strrchr(*argv, '/')))
	progname++;
   else
	progname = *argv;
   argv++;
   argc--;

   scheme = SCHEME_URLGET;
   if (!strcmp(progname, "httpget")) scheme = SCHEME_HTTPGET;
   if (!strcmp(progname, "ftpget")) scheme = SCHEME_FTPGET;
   if (!strcmp(progname, "ftpput")) scheme = SCHEME_FTPPUT;

   while (*argv[0] == '-') {
	switch (argv[0][1]) {
	case 'h':
		opt_h = -1;
		break;
	case 'v':
		opt_v = -1;
		break;
	case 'd':
		opt_d = -1;
		break;
	case 'p':
		opt_p = -1;
		break;
	case 'U':
		scheme = SCHEME_URLGET;
		break;
	case 'F':
		scheme = SCHEME_FTPGET;
		break;
	case 'P':
		scheme = SCHEME_FTPPUT;
		break;
	case 'H':
		scheme = SCHEME_HTTPGET;
		break;
	default:
		return usage();
	}
	argv++;
	argc--;
   }

   if (scheme == SCHEME_FTPGET || scheme == SCHEME_FTPPUT) {
	if (argc < 2 || argc > 5)
		return usage();

	/* FIXME: Add the ability to specify input file separately for ftpput */

   	strncpy(host, *argv++, sizeof(host));
	if ((p = strchr(host, ':'))) {
		*p++ = '\0';
		port = atoi(p);
	} else
   		port = 21;
   	path = *argv++;
   	if (argc && (*argv[0] != '-')) {
   		strncpy(user, *argv++, sizeof(user));
   		argc++;
   	} else
   		*user = '\0';
   	if (argc && (*argv[0] != '-')) {
   		strncpy(pass, *argv++, sizeof(pass));
   		argc++;
   	} else
   		*pass = '\0';
	if (scheme == SCHEME_FTPPUT) {
		type = 'S';		/* Always send files as binary, 'S' is the put vs. get flag */
		if (*argv[0] == '-')	/* Allow ftpput to use stdin */
			use_stdin++;
	}
	s = ftpio(host, port, user, pass, path, type, opt_v);
	return(s);
   }
   if (scheme == SCHEME_HTTPGET) {
	if (argc != 2)
		return usage();
   	strncpy(host, *argv++, sizeof(host));
	if ((p = strchr(host, ':'))) {
		*p++ = '\0';
		port = atoi(p);
	} else
   		port = 80;
   	path = *argv++;
	s = httpget(host, port, user, pass, path, opt_h, opt_d, opt_p);
	return(s);
   }

   if (argc != 1)
	return usage();

   url = *argv++;
   argc--;

   if (strncmp(url, "http://", 7) == 0) {
	scheme = SCHEME_HTTPGET;
   	ps = url + 7;
   } else
   if (strncmp(url, "ftp://", 6) == 0) {
	if (scheme != SCHEME_FTPPUT)
		scheme = SCHEME_FTPGET;
   	ps = url + 6;
   } else
   if (strncmp(url, "tcp://", 6) == 0) {
	scheme = SCHEME_URLGET;
   	ps = url + 6;
   } else {
	errmsg("Must specify http://, ftp:// or tcp:// url prefix");
	return(-1);
   }

   user[0] = '\0';
   pass[0] = '\0';
   host[0] = '\0';
   port = 0;

   p = ps;
   while (*p && *p != '/') p++;
   path = p;
   *path = '\0';

   at = strchr(ps, '@');
   if (at != (char *)NULL) {
   	*at = '\0';
   	p = ps;
   	while (*p && *p != ':') p++;
   	if (*p)
   		*p++ = '\0';
	strcpy(user, ps);
   	strcpy(pass, p);
   	ps = at + 1;
   }

   *path = '/';
   p = ps;
   while (*p && *p != '/' && *p != ':') p++;
   if (*p) {
   	strncpy(host, ps, p - ps);
   	host[p - ps] = '\0';
   }
   if (*p == ':') {
   	p++;
   	ps = p;
   	while (*p && *p != '/')
   		port = port * 10 + (*p++ - '0');
   }
   if (*p == '/')
	path = p;
   else
   	path = "/";
   if (scheme == SCHEME_FTPGET || scheme == SCHEME_FTPPUT) {
   	p = path;
   	while (*p && *p != ';') p++;
   	if (*p) {
   		*p++ = '\0';
   		if (strncasecmp(p, "type=", 5) == 0) {
   			p += 5;
   			type = tolower(*p);
   		}
   	}
   }

#if TRANS_DEBUG
   fprintf(stderr, "Host: %s\n", host);
   fprintf(stderr, "Port: %d\n", port);
   fprintf(stderr, "User: %s\n", user);
   fprintf(stderr, "Pass: %s\n", pass);
   fprintf(stderr, "Path: %s\n", path);
   fprintf(stderr, "Type: %c\n", type);
#endif

   switch(scheme) {
        case SCHEME_HTTPGET:
		s = httpget(host, port, user, pass, path, opt_h, opt_d, opt_p);
		break;
	case SCHEME_FTPGET:
	case SCHEME_FTPPUT:
		s = ftpio(host, port, user, pass, path, type, opt_v);
		break;
	case SCHEME_URLGET:
		s = tcpget(host, port, user, pass, path);
		break;
   }

   return(s);
}
